File: approval_engine/settlement/customization/supplier/supplier.js
// Copyright (c) 2026 8848 Digital LLP. All rights reserved.
// Proprietary and confidential. Unauthorized copying, distribution, or use
// of this file, via any medium, is strictly prohibited without prior
// written permission from 8848 Digital LLP.

frappe.ui.form.on("Supplier", {
	setup: function (frm) {
		frm.set_query("nature_of_service", function () {
			return {
				query: "approval_engine.settlement.api.v1.supplier.get_nature_of_service_query",
			};
		});
	},
	before_workflow_action: async function (frm) {
		const action = frm.selected_workflow_action;

		if (["Comment", "Update", "Reject", "Cancel"].includes(action)) {
			await handle_comment_or_reject(frm, action);
		} else if (action === "Approve") {
			await handle_approve_action(frm);
		}
	},
	validate: function (frm) {
		msa_agreement(frm);
	},
});

async function handle_comment_or_reject(frm, action) {
	frappe.dom.unfreeze();

	const label = action === "Reject" ? "Reason" : "Comment";

	return new Promise((resolve, reject) => {
		frappe.prompt(
			[
				{
					label: label,
					fieldname: "message",
					fieldtype: "Small Text",
					reqd: 1,
				},
			],
			(values) => {
				send_supplier_message(frm, values.message, action);
				resolve();
			},
			`Please provide ${label.toLowerCase()} for <b>${action}</b>:`,
			"Submit"
		);
	});
}

async function handle_approve_action(frm) {
	if (frm.doc.email_id) {
		send_approval_mail(frm);
	} else {
		frappe.msgprint(__("Supplier email not found"));
	}
}

function send_supplier_message(frm, message, action) {
	frappe.call({
		method: "approval_engine.settlement.api.v1.supplier.send_supplier_message",
		args: {
			supplier: frm.doc.name,
			reason: message,
			action: action,
		},
		callback: function (r) {
			if (r.data === "sent") {
				frappe.msgprint(__("Message sent to supplier successfully."));
			} else {
				frappe.msgprint(__("Supplier email not found."));
			}
		},
	});
}

function send_approval_mail(frm) {
	frappe.call({
		method: "approval_engine.settlement.api.v1.supplier.send_approval_mail",
		args: {
			supplier_name: frm.doc.supplier_name,
			email_id: frm.doc.email_id,
		},
		callback: function (r) {
			if (r.data === "sent") {
				frappe.msgprint(__("Approval mail sent to supplier successfully."));
			}
		},
	});
}

function msa_agreement(frm) {
	if (frm.doc.custom_msa_agreement === "Yes") {
		frm.set_df_property("custom_msa_agreement", "read_only", 1);
	} else {
		frm.set_df_property("custom_msa_agreement", "read_only", 0);
	}
}

File: approval_engine/settlement/api/v1/supplier.py
# Copyright (c) 2026 8848 Digital LLP. All rights reserved.
# Proprietary and confidential. Unauthorized copying, distribution, or use
# of this file, via any medium, is strictly prohibited without prior
# written permission from 8848 Digital LLP.
"""Whitelisted endpoints for Supplier customization (status mail, approval
mail, Nature of Service search).

Thin wrappers only -- the actual logic lives in
settlement/customization/supplier/utils.py and
settlement/customization/supplier/query.py.
"""

import frappe

from approval_engine.settlement.customization.supplier.query import get_nature_of_service_options
from approval_engine.settlement.customization.supplier.utils import (
	_send_supplier_status_mail,
	approval_mail,
)


@frappe.whitelist(methods=["POST"])
def send_supplier_message(supplier: str, reason: str, action: str):
	"""
	Notify a supplier by email of a Reject/Comment workflow action, and log
	the reason as a comment on the Supplier document.

	**Endpoint:** `/api/method/approval_engine.settlement.api.v1.supplier.send_supplier_message`
	**HTTP Method:** POST
	**Parameters:**
	        - supplier (str, required): The Supplier document name
	        - reason (str, required): The reason/comment text
	        - action (str, required): The workflow action ("Reject" or any other, treated as "Comment")
	**Response:** `"sent"` or `"no_email"`, serialized as JSON.
	"""
	return _send_supplier_status_mail(supplier, reason, action)


@frappe.whitelist(methods=["POST"])
def send_approval_mail(supplier_name: str, email_id: str):
	"""
	Notify a supplier by email that their registration has been approved.

	**Endpoint:** `/api/method/approval_engine.settlement.api.v1.supplier.send_approval_mail`
	**HTTP Method:** POST
	**Parameters:**
	        - supplier_name (str, required): The supplier's display name (used in the email body)
	        - email_id (str, required): The recipient email address
	**Response:** `"sent"` or `"no_email"`, serialized as JSON.
	"""
	return approval_mail(supplier_name, email_id)


@frappe.whitelist(methods=["GET", "POST"])
@frappe.validate_and_sanitize_search_inputs
def get_nature_of_service_query(
	doctype: str, txt: str, searchfield: str, start: int, page_len: int, filters: dict
):
	"""
	Returns all matching TDS Reference records (no pagination) so the
	Nature of Service Table MultiSelect on Supplier is not capped at
	the default page length of 20.

	**Endpoint:** `/api/method/approval_engine.settlement.api.v1.supplier.get_nature_of_service_query`
	**HTTP Method:** GET, POST
	**Parameters:**
	        - doctype (str, required): The Link/MultiSelect target doctype (standard search-query arg)
	        - txt (str, required): The text typed into the search box
	        - searchfield (str, required): The field being searched (standard search-query arg)
	        - start (int, required): Pagination offset (standard search-query arg, unused here)
	        - page_len (int, required): Page size (standard search-query arg, unused here)
	        - filters (dict, required): Additional filters (standard search-query arg, unused here)
	**Response:** List of (name, section_as_per_it_act_1961, tds_rate) tuples, serialized as JSON.
	"""
	return get_nature_of_service_options(txt)
